### PR TITLE
added a new project: Tic Tac Toe

### PR DESCRIPTION
**Pull request by joshi-kaushal**
# Description
I have added a new beginner level project: Tic Tac Toe.
It focuses on `useState` hook and cretes a two player tic tac toe game 

> Note: I have added three more packages: `react-toastify`, `react-icons` and `reactstrap`.

Fixes issue #60: TicTacToe

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
`iT wOrKs On My MaChIne`
nothing particular

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by atapas** [comment]
Hello @joshi-kaushal 

Thanks a lot for the PR. Could you please create an issue for this PR and link them?

**Comment by joshi-kaushal** [comment]
oh I totally forgot that, created a new issue, could you please take a look and assign it to me

**Comment by atapas** [comment]
> oh I totally forgot that, created a new issue, could you please take a look and assign it to me

Done.. Can we start reviewing it?

File: package.json
Comment (atapas) [inline package.json:9]: Do we really need bootstrap? As we want to be unbiased of any CSS framework like bootstrap, material, etc. as much as possible and encourage hands-on CSS as much.

In a long term, we will take call which one it will be. We may go with something like Tailwind also but subject to discussion. So if you could replace the Cards, Col Rows with simple Divs with flex/grid etc, would be great.

File: src/meta/play-meta.js
Comment (atapas) [inline src/meta/play-meta.js:126]: As you do not have a cover image URL, please drop a cover.png file in your play folder. Create some relevant image using tool like Canva.

File: src/plays/tic-tac-toe/Game.jsx
Comment (atapas) [inline src/plays/tic-tac-toe/Game.jsx:12]: Given a related comment about bootstrap

**Review by atapas — CHANGES_REQUESTED** [review]
@joshi-kaushal It is a great inclusion. Thanks for the play. I'm submitting my reviews.

File: package.json
Comment (joshi-kaushal) [reply package.json:9]: Okay, but it'd take me a few more days to convert everything to plain CSS.

If we decide to choose Tailwind in the future, I can add more projects from my learning days.

**Comment by atapas** [comment]
Hello @joshi-kaushal 

We have upgraded to React v18 as part of this issue: https://github.com/atapas/react-play/issues/87

We have tested the existing plays. All good. Please pull this change to your branch and keep it up-to-date. You have to do sanity of things to make sure everything works for your changes in your branch.

Happy Coding!

**Comment by atapas** [comment]
@joshi-kaushal  Once you done, please remove the work-in-progress label and review-ready label so that we can start the review.

File: src/plays/tic-tac-toe/TicTacToe.css
Comment (atapas) [inline src/plays/tic-tac-toe/TicTacToe.css:13]: Do not override selectors directly. Scope it to your play with a related class name.

**Review by atapas — CHANGES_REQUESTED** [review]
@joshi-kaushal Build fails...

File: src/meta/play-meta.js
Comment (joshi-kaushal) [reply src/meta/play-meta.js:5]: any idea what could be causing the error?

image

---

Here's `TicTacToe.jsx`

image

Comment (atapas) [reply src/meta/play-meta.js:5]: @joshi-kaushal Yes, this structure has been changed. Now the plays do not depends on the location to get id. Rather id passes as props from the Route controller. This fix was done to achieve the deep linking. Check the entry point of any other plays, you will see, we pass the id to to. Also, you can remove the references of useLocation.

I hope it helps. Please mark the conversation as closed when done.

**Comment by joshi-kaushal** [comment]
Hello @atapas, I have made required changes, please have a look.

File: src/plays/tic-tac-toe/TicTacToe.jsx
Comment (kirillkurko) [inline src/plays/tic-tac-toe/TicTacToe.jsx:9]: Why not to destructure id right here?

File: src/plays/tic-tac-toe/Game.css
Comment (kirillkurko) [inline src/plays/tic-tac-toe/Game.css:5]: Let's insert a newline here

File: src/plays/tic-tac-toe/Game.jsx
Comment (kirillkurko) [inline src/plays/tic-tac-toe/Game.jsx:15]: You can have a draw, so it's not really a win message, it will be better to go with something like finalMessage or endgameMessage

File: src/plays/tic-tac-toe/Game.jsx
Comment (kirillkurko) [inline src/plays/tic-tac-toe/Game.jsx:19]: This is a bad idea to have a component inside component cuz you will have a new instance of this function on every re-render. It's better to move it to a separate component and pass needed props (here you just need to pass onClick handler)

Comment (kirillkurko) [inline src/plays/tic-tac-toe/Game.jsx:19]: It's a button, so let's name it as a PlayAgainButton to make it clear

File: src/plays/tic-tac-toe/Game.jsx
Comment (kirillkurko) [inline src/plays/tic-tac-toe/Game.jsx:21]: Component shouldn't know about it's positioning, let's remove this div (finally button is just a button and components that use this button should decide how to place it)

File: src/plays/tic-tac-toe/Game.jsx
Comment (kirillkurko) [inline src/plays/tic-tac-toe/Game.jsx:137]: Here you will have `<div className="center"><PlayAgainButton onClick={onPlayAgainClick}></div>`

File: src/plays/tic-tac-toe/TicTacToe.jsx
Comment (kirillkurko) [inline src/plays/tic-tac-toe/TicTacToe.jsx:5]: Do we really need this comment? I think import is saying what are you doing here, so comment is useless

File: src/plays/tic-tac-toe/TicTacToe.css
Comment (kirillkurko) [inline src/plays/tic-tac-toe/TicTacToe.css:12]: Let's insert a newline here

File: src/plays/tic-tac-toe/Icon.js
Comment (kirillkurko) [inline src/plays/tic-tac-toe/Icon.js:11]: Let's maybe return null or undefined as a default case, cuz from Icon you expect to receive a ReactElement or nothing, but not an empty string

**Review by kirillkurko — COMMENTED** [review]
I'm not going to request change, cuz I'm not sure I can do this here, but I left comments that I think should be addressed

**Comment by atapas** [comment]
> I'm not going to request change, cuz I'm not sure I can do this here, but I left comments that I think should be addressed

You can surely. That's the fun of ReactPlay, we all learn together 🙌

**Comment by joshi-kaushal** [comment]
> I'm not going to request change, cuz I'm not sure I can do this here, but I left comments that I think should be addressed

Hey, thanks man! I really appreciate this. I had built this in my initial learning days, hence so much spaghetti code. I will make these changes ASAP. 

Thanks again, looking forward fo more reviews :)

**Comment by kirillkurko** [comment]
> Hey, thanks man! I really appreciate this. I had built this in my initial learning days, hence so much spaghetti code. I will make these changes ASAP.
> 
> Thanks again, looking forward fo more reviews :)

I see, cool, I will try to be helpful with my reviews

**Comment by kirillkurko** [comment]
> You can surely. That's the fun of ReactPlay, we all learn together 🙌

Cool, I will try to become a reviewer

**Comment by atapas** [comment]
@joshi-kaushal Once you taken care of the comments, please resolve them by saying done. This will help me to do a final review once all the existing comments are done. Looking forward to it.

**Comment by joshi-kaushal** [comment]
I have a free time after my exam of 18th. I will finish remaining tasks later.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by atapas** [comment]
> I have a free time after my exam of 18th. I will finish remaining tasks later. 

Sounds perfect. All the best for the exam.

**Comment by atapas** [comment]
Hello @joshi-kaushal 

We will be moving the react-play project from the `atapas` account to an organization to help manage the project in a better way. The new organization URL is: https://github.com/reactplay.

> **The movement will take place between 12:30 - 14:00 hrs IST(7:00 am - 8:30 am GMT), 30th May 2022. Please refrain from pushing any changes between this time.**

We will inform you when the movement is complete. You need to make a few minimal configuration changes after that. We will let you know.

**Comment by joshi-kaushal** [comment]
@atapas 

Sure, keep me updated. I haven't been able to make changes since the last time we discussed so everything must be clear from my side.

**Comment by atapas** [comment]
@joshi-kaushal 

- Pull the main
- Resolve merge conflicts
- COmment on each of the comments mentioning "done" or not done.


Lets close this one asap.

File: package.json
Comment (joshi-kaushal) [reply package.json]: done

File: src/meta/play-meta.js
Comment (joshi-kaushal) [reply src/meta/play-meta.js]: done

File: src/plays/tic-tac-toe/Game.jsx
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/Game.jsx]: done

File: src/plays/tic-tac-toe/TicTacToe.jsx
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/TicTacToe.jsx]: done

File: src/plays/tic-tac-toe/Game.jsx
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/Game.jsx]: done

File: src/plays/tic-tac-toe/README.md
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/README.md]: done

File: src/plays/tic-tac-toe/TicTacToe.css
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/TicTacToe.css]: done

File: src/meta/play-meta.js
Comment (joshi-kaushal) [reply src/meta/play-meta.js:5]: done

File: src/plays/tic-tac-toe/TicTacToe.jsx
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/TicTacToe.jsx:8]: done

File: src/plays/tic-tac-toe/TicTacToe.css
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/TicTacToe.css:12]: done

**Comment by joshi-kaushal** [comment]
@atapas everything is clear from my side, ready for final review!

File: src/plays/tic-tac-toe/TicTacToe.jsx
Comment (atapas) [inline src/plays/tic-tac-toe/TicTacToe.jsx:20]: Add a unique class name here and use that to prefix all the styles so that no chance of any conflicts.

**Review by atapas — CHANGES_REQUESTED** [review]

File: src/plays/tic-tac-toe/TicTacToe.jsx
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/TicTacToe.jsx:20]: I can move `tictactoe.css` to `Game.jsx` and it would work the same. Can I do that?


Or, are you suggesting that I need to rename every CSS class from the entire play with a unique prefix? 
Like, `ttt_buttonGroup`, `ttt_icon` etc?

File: src/plays/tic-tac-toe/Game.css
Comment (koustov) [inline src/plays/tic-tac-toe/Game.css:1]: I would suggest using a more specific class name. Something like `tic-tac-tow-grid`

File: yarn.lock
Comment (koustov) [inline yarn.lock:1]: Do not check in lock files.

**Review by koustov — CHANGES_REQUESTED** [review]
Few review comments added

File: src/plays/tic-tac-toe/Game.css
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/Game.css:35]: It's working properly. Still, should i change/remove this?

File: yarn.lock
Comment (joshi-kaushal) [reply yarn.lock:1]: Should I add it in `.gitignore`?

File: src/plays/tic-tac-toe/Game.css
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/Game.css:1]: replaced it with `ttc_grid`.

File: src/plays/tic-tac-toe/Game.jsx
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/Game.jsx]: Thanks! This is really impressive.

I made some changes, it is working.

File: yarn.lock
Comment (atapas) [reply yarn.lock:1]: It's already added. You may have to get latest from main.

**Comment by koustov** [comment]
@joshi-kaushal please pull the main and resolve conflicts. Also let us know if it is review ready?

**Comment by joshi-kaushal** [comment]
@koustov everything is clear from my side.

**Review by joshi-kaushal — COMMENTED** [review]
changes done

**Comment by koustov** [comment]
> @koustov everything is clear from my side.

@joshi-kaushal I see some comments are addressed however few others are not. These are just marked as `Resolved`. So I am not sure what is the fate of those comments.

I would suggest, if you don't agree to any comment, give a proper justification and then mark resolved. 
Similar way, if a comment has been addressed, at minimum one should reply by saying `done` and resolve it.

File: src/plays/tic-tac-toe/TicTacToe.css
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/TicTacToe.css:19]: I realized it is unnecessary, hence removing it completely. Must be something I did white building and forgot to delete.

File: src/plays/tic-tac-toe/TicTacToe.css
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/TicTacToe.css]: done.

File: src/plays/tic-tac-toe/Game.css
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/Game.css:19]: added flexbox, it's centred aligned now.

File: src/plays/tic-tac-toe/Game.css
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/Game.css]: I reduced the opacity to `0.75`.

File: src/plays/tic-tac-toe/Game.jsx
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/Game.jsx:56]: made these changes.

File: src/plays/tic-tac-toe/Game.jsx
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/Game.jsx:20]: added a new state: `isFinished`.

**Comment by joshi-kaushal** [comment]
I forgot to push one change but other than that I believe I have made every change you asked for. I don't reply **done** everywhere to avoid spam.

I will keep these things in mind the next time.

If there is still some comment that I have resolved but haven't acknowledged, I have done it by mistake.

File: src/plays/tic-tac-toe/Game.css
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/Game.css]: mad it mobile responsive.

File: src/plays/tic-tac-toe/Game.css
Comment (joshi-kaushal) [reply src/plays/tic-tac-toe/Game.css:8]: Forgot to do this earlier, marked it resolved without actually resolving it. My bad.

**Comment by atapas** [comment]
Hi @joshi-kaushal 

We have some massive changes in the data store and we are changing how the play gets created for a better usability. As this play is interim, we need to make sure it migrates correctly. Could you please ping me on Discord so that I make sure everything is placed?

https://discord.gg/YbmXwU5f

**Comment by koustov** [comment]
Hey @joshi-kaushal once you are done with merge-conflict, please notify us on [👍-play-review](https://discord.com/channels/982239924227031070/987806556445892618) channel

**Comment by joshi-kaushal** [comment]
@koustov I am ill. it will take me three or four days to do this. I hope that's fine.

**Comment by koustov** [comment]
> @koustov I am ill. it will take me three or four days to do this. I hope that's fine.

no problem. health first. take care

**Comment by joshi-kaushal** [comment]
@koustov please have a look.

**Comment by joshi-kaushal** [comment]
Sure. My current schedule is occupied with offline exams so it might take
some time.

**Comment by koustov** [comment]
I am good with the PR however getting following error while deploying 
image

**Comment by joshi-kaushal** [comment]
I fetched from upstream, and the play shows someone else's name as a creator. Could you please help me here?
image

**Comment by joshi-kaushal** [comment]
Also the error was about `package.json`. It should be fixed now.

**Comment by koustov** [comment]
> I fetched from upstream, and the play shows someone else's name as a creator. Could you please help me here? image

Ok, I will check and update

**Comment by joshi-kaushal** [comment]
@koustov did you get time to check?

**Comment by joshi-kaushal** [comment]
@koustov did you get time to check?

**Comment by koustov** [comment]
@joshi-kaushal  I am ready to merge this one. Please resolve conflict and notify me

**Review by koustov — DISMISSED** [review]
lgtm

**Comment by atapas** [comment]
@joshi-kaushal 

image

**Comment by joshi-kaushal** [comment]
@atapas please check now

**Comment by joshi-kaushal** [comment]
@atapas can we wind this up please?

**Comment by atapas** [comment]
@joshi-kaushal Any update?

**Comment by atapas** [comment]
@joshi-kaushal 

Please check the build error

image

File: package.json
Comment (atapas) [inline package.json:71]: @joshi-kaushal Take care of this.

**Review by atapas — CHANGES_REQUESTED** [review]

**Comment by atapas** [comment]
@koustov do we need this? Can you please take action accordingly?

**Pull request closed without merging**
